Util/src/main/java/io/deephaven/util/datastructures/LazyCachingFunction.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.util.datastructures;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.ref.SoftReference;
+import java.util.function.Function;
+
+/**
+ * {@link Function} wrapper that caches the result in a {@link SoftReference}. Only suitable to wrap suppliers that are
+ * safely repeatable and don't return {@code null}.
+ */
+public final class LazyCachingFunction<INPUT_TYPE, OUTPUT_TYPE> implements Function<INPUT_TYPE, OUTPUT_TYPE> {
+
+    private final Function<INPUT_TYPE, OUTPUT_TYPE> internalSupplier;
+
+    private volatile SoftReference<OUTPUT_TYPE> cachedResultRef;
```

Comment:
Don't we need to use a Map to cache? What happens if apply is provided w/ a different arg?

Reply:
The dictionary data is to be read from the underlying parquet file. 
So irrespective of the context, we will always read the same data since the underlying file is the same.